content/docs/user-guide/dvc-files.md

```diff
@@ -237,14 +240,227 @@ stages:
+grouped by parameters file. And in the case of templated/parameterized
+`dvc.yaml` files (see next section), their actual values are substituted into
+the `dvc.lock` YAML structure.
+
+### Parameterize `dvc.yaml`
+
+`dvc.yaml` supports a templating format to reuse values from different sources
+in the YAML structure itself. The sources can be external
+[parameters](/doc/command-reference/params) files, or internal `vars` sections,
+as detailed below.
+
+Let's say we have `params.yaml` (default params file) with the following
+contents:
+
+```yaml
+models:
+  us:
+    threshold: 10
+    filename: 'model-us.hdf5'
+```
+
+Those values can be used anywhere in `dvc.yaml` with the `${}` _substitution
+expression_:
+
+<!-- prettier-ignore-start -->
+```yaml
+stages:
+  build-us:
+    cmd: >-
+      python train.py
+      --tresh ${models.us.threshold}
+      --out ${models.us.filename}
+    outs:
+      - ${models.us.filename}:
+        cache: true
+```
+<!-- prettier-ignore-end -->
+
+Alternatively, these or other values could be listed as top-level `vars` like
+this:
+
+```yaml
+vars:
+  - models:
+      us:
+        threshold: 10
+  - desc: 'Reusable description'
+
+stages:
+  build-us:
+    desc: ${desc}
+    cmd: python train.py --tresh ${models.us.threshold} ...
+```
+
+⚠️ Both external params and top `vars` can coexist as long as their contents
+don't overlap (so the 2 examples above are incompatible with each other).
+
+To load additional params files, list them in the top `vars` section in the
+desired order, e.g.:
+
+```yaml
+vars:
+  - params.json
+  - myvar: 'value'
+  - myconfig.yaml
+```
+
+> Note that the default `params.yaml` is always included first.
+
+It's also possible to specify what to include from extra params files, with a
+`:` colon:
+
+```yaml
+vars:
+  - params.json:cleaning,featurization
+
+stages:
+  clean:
+    cmd: source ${cleaning.script}
+    outs:
+      - ${cleaning.outname}
+  featurize:
+    cmd: python ${featurization.pyscript}
+    deps:
+      - ${cleaning.outname}
+    outs: ...
+```
+
+⚠️ DVC merges data from params files as long as there are no leaf node
+collisions. For example `{"grp": {"a": 1}}` can be merged with
+`{"grp": {"b": 2}}`, but not with `{"grp": {"a": 7}}`.
+
+Stage-specific values are also supported, with inner `vars` sections:
+
+```yaml
+stages:
+  build-us:
+    vars:
+      - model:
+        filename: 'model-us.hdf5'
+    cmd: python train.py --out ${model.filename}
+    outs:
+      - ${model.filename}
+```
+
+The substitution expression supports these forms:
+
+```yaml
+${param} # Simple
+${param.key} # Nested values through . (period)
+${param.list[0]} # List elements via index in [] (square brackets)
+```
+
+> To use the expression literally in `dvc.yaml`, escape it with a backslash,
+> e.g. `\${...`.
+
+### Stage groups
+
+You can generate more than one stage from a single `dvc.yaml` entry with the
+following syntax. A `foreach` element accepts a list or dictionary with values
```

Comment:
### Terminology discussion

Summary (from https://github.com/iterative/dvc.org/pull/2052#pullrequestreview-558972873, https://github.com/iterative/dvc.org/pull/2052#pullrequestreview-558972942, iterative/dvc/issues/5165):

> Better to just say foreach or foreach loops?
> we'll need a proper document which will connect dots and explain these new concepts from the user perspective... implementing this based on a lot of tickets - how did users ask about this? what terminology
> loop structures sounds too low level
> They are not a group in any way
> A group is "more than one of the same thing, regarded as a unit"
> in our case it's not a group, these stages are not different from any other stages
> On the dvc.yaml level it's indeed looks like a single "entity"

Takeaways:

- we don't like group, because group is a very specific term in math/languages
- totally fine to use "group" in some informal way. The misleading would be to make it formal (use in title
- @skshetry can you maybe help me find these original tickets or requests from users to check on that? Thanks

Reply:
For now I'm using "Defining multiple stages at once" as title and introduced the term "multi-stage".